xdsl/frontend/jaxpr/__init__.py

```diff
@@ -1,17 +1,38 @@
-from dataclasses import dataclass
+from dataclasses import dataclass, field
+from typing import NoReturn
 
 from jax._src.core import ClosedJaxpr
 
-from xdsl.builder import Builder
+from xdsl.builder import Builder, ImplicitBuilder
 from xdsl.dialects.builtin import FunctionType, ModuleOp, TensorType, f32
 from xdsl.dialects.func import FuncOp, Return
-from xdsl.ir import Block, Region
+from xdsl.ir import Block, Region, SSAValue
+
+from .primitive_builder import JaxPrimitiveBuilder
 
 
 class IRGenError(Exception):
     pass
 
 
+@dataclass
+class ScopedSymbolTable:
+    "A mapping from variable names to SSAValues, append-only"
+
+    table: dict[str, SSAValue] = field(default_factory=dict)
+
+    def __contains__(self, __o: object) -> bool:
+        return __o in self.table
+
+    def __getitem__(self, __key: str) -> SSAValue:
+        return self.table[__key]
+
+    def __setitem__(self, __key: str, __value: SSAValue) -> None:
+        if __key in self:
+            raise AssertionError(f"Cannot add value for key {__key} in scope {self}")
+        self.table[__key] = __value
```

Comment:
if we're reusing this helper from Toy, could you please first move this into `utils` in a separate PR, add some tests, and then use that version here? It feels like something we'll be reusing in the future, and MLIR has a version for clients to use.

Reply:
I've started doing this locally, as I'd like it to be in a specific place, and potentially reused in the interpreter. Will open a PR today.

Reply:
#3355